packages: set Epoch of 1 if package version less than Bottlerocket release version

ginglis13 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**


Related: 

* https://github.com/bottlerocket-os/twoliter/issues/364
* https://github.com/bottlerocket-os/bottlerocket/discussions/4063

**Description of changes:**

For all packages whose version is < 3.0.0, set an Epoch of 1 in the package spec. This is to appease new app inventory requirements for building Bottlerocket variants such that package version comparisons using rpmvercmp are not broken.

See https://github.com/bottlerocket-os/bottlerocket/discussions/4063 for full details

1. Get App Inventory for aws-dev, metal-dev, vmware-dev and combine

Build with Twoliter from https://github.com/bottlerocket-os/twoliter/pull/384

```
for variant in aws-dev metal-dev vmware-dev; do
        cargo make \
                   -e BUILDSYS_ARCH=x86_64 \
                   -e=TWOLITER_REPO=https://github.com/ginglis13/twoliter \
                  -e=TWOLITER_VERSION=44a2623a6419dc7b07e346175a425bcad2f88aa6 \
                  -e=TWOLITER_ALLOW_SOURCE_INSTALL=true \
                  -e=TWOLITER_ALLOW_BINARY_INSTALL=false \
                  -e=TWOLITER_SKIP_VERSION_CHECK=true \
                  -e BUILDSYS_VARIANT=$variant
done
 
# jq for just package name, version, epoch
cat build/images/x86_64-aws-dev/latest/application-inventory.json | jq -r '.Content[] | [.Name, .Version, .Epoch] | @csv' \
    > aws-dev-latest.csv
    
cat build/images/x86_64-metal-dev/latest/application-inventory.json | jq -r '.Content[] | [.Name, .Version, .Epoch] | @csv' \
    > metal-dev-latest.csv
    
cat build/images/x86_64-vmware-dev/latest/application-inventory.json | jq -r '.Content[] | [.Name, .Version, .Epoch] | @csv' \
    > vmware-dev-latest.csv
    
# combine unique to one csv
sort -u *.csv -o bottlerocket-packages.csv

# strip quotes
  sed -i 's/\"//g' bottlerocket-packages.csv
```


3. Search the CSV for package,version for < 1.24.1

```
#!/usr/bin/env bash
# find_packages_requiring_epoch.sh

BOTTLEROCKET_VERSION=1.24.1

declare i_need_epoch=()

while IFS=, read -r name version epoch
do
        rpmdev-vercmp $epoch:$version:0 0:$BOTTLEROCKET_VERSION:0 > /dev/null
        if [ $? == 12 ]; then
                i_need_epoch+=("$name $version")
        fi

done < bottlerocket-packages.csv

printf '%s\n' "${i_need_epoch[@]}"
```

```
apiclient 0.0
apiserver 0.0
bloodhound 0.0
bootstrap-commands 0.0
bootstrap-containers 0.0
bork 0.0
bottlerocket-metadata 1.0
bottlerocket-settings-defaults 0.0
bottlerocket-settings-defaults-aws-dev 0.0
bottlerocket-settings-defaults-metal-dev 0.0
bottlerocket-settings-defaults-vmware-dev 0.0
bottlerocket-settings-plugins 0.0
bottlerocket-settings-plugins-aws-dev 0.0
bottlerocket-settings-plugins-metal-dev 0.0
bottlerocket-settings-plugins-vmware-dev 0.0
certdog 0.0                                                                                                                                                                                                                                                                                                         [0/589]
cfsignal 0.0
conntrack-tools 1.4.8
containerd 1.7.22
containerd-bin 1.7.22
corndog 0.0
early-boot-config 0.1
early-boot-config-aws 0.1
early-boot-config-local 0.1
early-boot-config-metal 0.1
early-boot-config-vmware 0.1
filesystem 1.0
ghostdog 0.0
host-containers 0.0
host-ctr 0.0
host-ctr-bin 0.0
hostname-imds 0.1.1
hostname-reverse-dns 0.1.1
iptables 1.8.10
keyutils 1.6.1
libelf 0.191
libgcc 0.0
libmnl 1.0.5
libnetfilter_conntrack 1.0.9
libnetfilter_cthelper 1.0.1
libnetfilter_cttimeout 1.0.1
libnetfilter_queue 1.0.5
libnfnetlink 1.0.2
libnftnl 1.2.7
libnvme 1.10
libstd-rust 0.0
libtirpc 1.3.5
liburcu 0.14.0
libz 1.3.1
libzstd 1.5.6
logdog 0.0
login 0.0.1
makedumpfile 1.7.5
metricdog 0.0
microcode-amd-license 0.0
microcode-intel-license 0.0
microcode-licenses 0.0
migration 0.0
netdog 0.1.1
netdog-systemd-networkd 0.1.1
oci-add-hooks 1.0.0
os 0.0
prairiedog 0.0
release 0.0
runc 1.1.14
runc-bin 1.1.14
schnauzer 0.0
selinux-policy 0.0
settings-committer 0.0
shibaken 0.0
shimpei 0.0
signpost 0.0
socat 1.8.0.0
soci-snapshotter 0.7.0
soci-snapshotter-bin 0.7.0
storewolf 0.0
sundog 0.0
thar-be-settings 0.0
thar-be-updates 0.0
updog 0.0
xfscli 0.0
```

Set `Epoch: 1` for all specs that build above packages

**Testing done:**

1. Publish core kit with these changes
2. Repeat (1) and (2) above, confirm Epoch is 1 for all packages with version < 1.24.1

The following packages are defined in the first party Bottlerocket AMI and are still showing as an Epoch of 0:

```
$ ./find_packages_requiring_epoch.sh
bottlerocket-metadata 1.0
bottlerocket-settings-defaults 0.0
bottlerocket-settings-defaults-aws-dev 0.0
bottlerocket-settings-defaults-metal-dev 0.0
bottlerocket-settings-defaults-vmware-dev 0.0
bottlerocket-settings-plugins 0.0
bottlerocket-settings-plugins-aws-dev 0.0
bottlerocket-settings-plugins-metal-dev 0.0
bottlerocket-settings-plugins-vmware-dev 0.0
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ginglis13 committed:
packages: set Epoch of 1 if package version < core kit release version

For all packages whose version is < 4.0.0, set an Epoch of 1 in the
package spec. This is to appease new app inventory requirements for
building Bottlerocket variants such that package version comparisons
using rpmvercmp are not broken.

See
https://github.com/bottlerocket-os/bottlerocket/discussions/4063 for
full details

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 commented:
^ set `Epoch: 1` for `early-boot-config`

ginglis13 commented:
^ Add more packages after using `rpmdev-vercmp`:

```
    modified:   packages/conntrack-tools/conntrack-tools.spec
    modified:   packages/containerd/containerd.spec
    modified:   packages/iptables/iptables.spec
    modified:   packages/keyutils/keyutils.spec
    modified:   packages/libtirpc/libtirpc.spec
    modified:   packages/libz/libz.spec
    modified:   packages/libzstd/libzstd.spec
    modified:   packages/makedumpfile/makedumpfile.spec
    modified:   packages/socat/socat.spec
    modified:   packages/wicked/wicked.spec
```

ginglis13 commented:
Need add an epoch to the following packages, given we should be comparing against bottlerocket-core-kit versions instead. I chose `3.0.0` conservatively (the core-kit's latest release is 2.8.4):

```
acpid 2.0.34
e2fsprogs 1.47.1
e2fsprogs-libs 1.47.1
glibc 2.38
grub 2.06
kexec-tools 2.0.29
libacl 2.3.2
libattr 2.5.2
libblkid 2.40.2
libcap 2.70
libexpat 2.6.3
libfdisk 2.40.2
libglib 2.78.4
libmount 2.40.2
libseccomp 2.5.5
libsmartcols 2.40.2
libuuid 2.40.2
nvme-cli 2.10.2
pigz 2.8
util-linux 2.40.2
```

and all the k8s packages.

bcressey commented:
> I chose 3.0.0 conservatively (the core-kit's latest release is 2.8.4)

We're going to hit 3.x soon when we remove the k8s 1.23 packages.

ginglis13 commented:
latest force push conservatively compares against a theoretical core-kit 4.0.0:

```
        modified:   packages/acpid/acpid.spec
        modified:   packages/e2fsprogs/e2fsprogs.spec
        modified:   packages/glibc/glibc.spec
        modified:   packages/grep/grep.spec
        modified:   packages/grub/grub.spec
        modified:   packages/kexec-tools/kexec-tools.spec
        modified:   packages/kubernetes-1.23/kubernetes-1.23.spec
        modified:   packages/kubernetes-1.24/kubernetes-1.24.spec
        modified:   packages/kubernetes-1.25/kubernetes-1.25.spec
        modified:   packages/kubernetes-1.26/kubernetes-1.26.spec
        modified:   packages/kubernetes-1.27/kubernetes-1.27.spec
        modified:   packages/kubernetes-1.28/kubernetes-1.28.spec
        modified:   packages/kubernetes-1.29/kubernetes-1.29.spec
        modified:   packages/kubernetes-1.30/kubernetes-1.30.spec
        modified:   packages/kubernetes-1.31/kubernetes-1.31.spec
        modified:   packages/libacl/libacl.spec
        modified:   packages/libattr/libattr.spec
        modified:   packages/libaudit/libaudit.spec
        modified:   packages/libcap/libcap.spec
        modified:   packages/libexpat/libexpat.spec
        modified:   packages/libffi/libffi.spec
        modified:   packages/libglib/libglib.spec
        modified:   packages/libseccomp/libseccomp.spec
        modified:   packages/libselinux/libselinux.spec
        modified:   packages/libsemanage/libsemanage.spec
        modified:   packages/libsepol/libsepol.spec
        modified:   packages/nvme-cli/nvme-cli.spec
        modified:   packages/pciutils/pciutils.spec
        modified:   packages/pigz/pigz.spec
        modified:   packages/policycoreutils/policycoreutils.spec
        modified:   packages/util-linux/util-linux.spec
```

ginglis13 committed:
packages: normalize on Release: 1

Set Release: 1%{?dist} for all packages to normalize.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>
